KEDamage.cs

```diff
@@ -8,58 +8,94 @@ class KEDamage:Damage
+            /// </summary>
+            public float HealthDamageFactor;
+            /// <summary>
+            /// Air friction constant used in calculation of attenuation
+            /// </summary>
             public float FrictionFactor;
         }
 
-        private KineticData data;
-        private Damage.Target target;
+        private KineticData _keData;
 
-        public KEDamage(KineticData d,Damage.Target t):base(EDamageTypes.KE)
+        public KEDamage(KineticData data,Target target)
+            : base(DamageTypes.KE, target)
         {
-            data = d;
-            target = t;
+            this._keData = data;
         }
 
-        override public float DealDamage()
+        public override Target CalculateDamage()
         {
+            Target finalState = this.CurrentTarget;
+            KineticData ke = _keData;
+
             // Calculate attenuation of air friction
-            data.Pierce = CalculateKEAttenuationSimple(data.Pierce,data.Distance,data.FrictionFactor);
+            ke.Pierce = CalculateKEAttenuationSimple(
+                ke.Pierce,
+                ke.Distance,
+                ke.FrictionFactor
+            );
 
-            // Calculate effects of ERA
-            data.Pierce = CalculatePostERAPierce(data.Pierce,target.ERAData.KEFractionMultiplier);
+            if (finalState.EraData.Value > 0.0f) {
+                // Calculate effects of ERA
+                float finalEra = Math.Max(
+                    0.0f,
+                    finalState.EraData.Value - ke.Pierce * finalState.EraData.KEFractionMultiplier
+                );
+                finalState.EraData.Value = finalEra;
+
+                ke.Pierce = CalculatePostERAPierce(
+                    ke.Pierce,
+                    finalState.EraData.KEFractionMultiplier
+                );
+            }
+            
+            
+
+            // Armor degradation
+            float finalArmor = Math.Max(
+                0.0f,
+                finalState.Armor - (ke.Pierce / finalState.Armor) * ke.Degradation
+            );
+            finalState.Armor = finalArmor;
 
             // Calculate final damage
-            float finalDamage = Math.Max(0.0f, target.Armor - data.Pierce)*data.HealthKEFactor;
-            return finalDamage;
+            float finalDamage = Math.Max(
+                0.0f,
+                (ke.Pierce - finalState.Armor) * ke.HealthDamageFactor
+            );
+            float finalHealth = Math.Max(
+                0.0f,
+                finalState.Health - finalDamage
+            );
+            finalState.Health = finalHealth;
+            
+            return finalState;
         }
 
-        // Then we implement a separate function for armour degradation,
-        // use callback function to update target armour value in its data model class
-
-        #region DamageCalculation
 
         private static float CalculateKEAttenuationSimple(float pierce, float distance, float frictionFactor)
```

Comment:
I'm certain you created a separate function for clarity but otherwise we would want to keep calculations like this inline with the other code and not separated.

Reply:
Ah it is because we might change this function later on (you know, maybe to a polynomial approximation), I made it a single function so that we could locate it more easily.